File: C/Computer Systems/bitwise/operations.c
#include <stdio.h>

void bitwise_not(int p);
void bitwise_and(int p, int q);
void bitwise_or(int p, int q);
void bitwise_xor(int p, int q);

int main(void) {
  int a = 0x4E;
  int b = 0xE1;

  bitwise_not(a);
  bitwise_not(b);

  bitwise_and(a, b);

  bitwise_or(a, b);

  bitwise_xor(a, b);

  return 0;
}

void bitwise_not(int p) {
  int r = ~p;
  printf("~%.2x = %.2x\n", p, r);
}

void bitwise_and(int p, int q) {
  int r = p & q;
  printf("%.2x & %.2x = %.2x\n", p, q, r);
}

void bitwise_or(int p, int q) {
  int r = p | q;
  printf("%.2x | %.2x = %.2x\n", p, q, r);
}

void bitwise_xor(int p, int q) {
  int r = p ^ q;
  printf("%.2x ^ %.2x = %.2x\n", p, q, r);
}

File: Rust/Computer Systems/computer_systems/src/bitwise/operations.rs
pub fn bitwise_not(p: u32) {
    println!("~{:02x} = {:02x}", p, !p);
}

pub fn bitwise_and(p: u32, q: u32) {
    println!("{:02x} & {:02x} = {:02x}", p, q, p & q);
}

pub fn bitwise_or(p: u32, q: u32) {
    println!("{:02x} | {:02x} = {:02x}", p, q, p | q);
}

pub fn bitwise_xor(p: u32, q: u32) {
    println!("{:02x} ^ {:02x} = {:02x}", p, q, p ^ q);
}
